🐛Fix for files not being deleted from cache

notanengineercom opened the pull request:
Fixes #10174 
This is an upgrade from #10175 @gargol @kirrg001 
And for the unzipped theme deletion on failure, I opened a [PR in the gscan repo](https://github.com/TryGhost/gscan/pull/172), as I explained there, it makes more sense to me to have the fix added to gscan.
***
This PR additions:
- [x] Detects when a response ends
- [x]  Uses a middleware
- [x] Deletes req.file.path
- [x] Deletes req.files[...].path 

Because some files are deleted manually in Ghost, the unlink fails sometimes and triggers an error (tries to delete a non existing path). This isn't an issue for me, but maybe some people may find that annoying? I don't know. 

So I don't know what to do here. Ignore it, leaving it as it is or maybe checking if the path exists before calling unlink, or even removing the manual deletion of files written before in the Ghost core? What do you think is the best solution?

Edited:
-------
Another question I have, should I add the middleware to the v0.1 API? It's just on the v2 right now.

gargol commented:
Hey @notanengineercom :wave: Nice work! Will go through raised points one by one :)

> the unlink fails sometimes and triggers an error

We should definitely try to avoid raising errors in contexts that don't need them. In this case, there are a couple of things to consider:
1. As you pointed out: "even removing the manual deletion of files written before in the Ghost core". If we enable this middleware, should we even be doing `fs.unlink` elsewhere in the related code anymore? If we introduce this central place that takes care of it, I think we don't need to worry about cleanup elsewhere.
2. If the file doesn't exist then we probably should log a warning, that would mean that we didn't follow the pt.1 or something else happened during handling the request and actually needs attention.

> should I add the middleware to the v0.1 API?

The v0.1 is not being developed anymore, so would rather avoid doing unnecessary changes there :)

---

Separate comment around enabling this middleware globally. Adding file removal to every single request in API v2 isn't the safest approach. As this is meant only to do clean up after multer uploads only, would apply it in a manner symmetric to where multer is [used](https://github.com/TryGhost/Ghost/blob/master/core/server/web/api/v2/admin/routes.js#L52-L57), this would become: 
```javascript
    router.post('/settings/routes/yaml',
        mw.authAdminApi,
        upload.single('routes'),
        shared.middlewares.uploadClear,
        shared.middlewares.validation.upload({type: 'routes'}),
        apiv2.http(apiv2.settings.upload)
    );
```

gargol commented on core/server/web/shared/middlewares/cache-clear.js:
```diff
@@ -0,0 +1,32 @@
+
+const deleteSingleFile = file => fs.unlink(file.path).catch(err => common.logging.error(err));
+
+const cacheClear = (req, res, next) => {
```
Not sure about naming here (and module file). `cache` is something that is used repeatedly, and what we are dealing here with is just a temporary file storage. What do you think about `uploadClear` or any other ideas?

gargol commented on core/server/web/shared/middlewares/cache-clear.js:
```diff
@@ -0,0 +1,32 @@
+const deleteSingleFile = file => fs.unlink(file.path).catch(err => common.logging.error(err));
+
+const cacheClear = (req, res, next) => {
+    const delFiles = () => {
```
The rule we are trying to keep is preferring long form variable naming to short ones. So, to be consistent with a method name above `deleteSingleFile` would go with  `deleteFiles` here :+1:

gargol commented:
@kirrg001 there were couple good points you raised in https://github.com/TryGhost/Ghost/pull/10175#issuecomment-443855956. This PR addresses: `register res.on event`. Could you please expand on the other two:
> - fs.unlink if not disabled
> - proper Ghost error handling

By `if not disabled` did you mean that we make this middleware optional?

kirrg001 commented:
> fs.unlink if not disabled

Hmm I don't remember 🤔 🤪Maybe i saw a case where a request/route might want to disable the behaviour?

notanengineercom commented:
@gargol thanks for all the feedback and suggestions.

> We should definitely try to avoid raising errors in contexts that don't need them. In this case, there are a couple of things to consider:

A. I thought that the logical thing to do is removing the lines of code in the core where `fs.unlink` is called to delete the files of multer, because right now they are not being deleted on all scenarios (only on success). So when the middleware is present, it should handle this task (that's why we are implementing it and if there are future routes for new type of files being uploaded with multer,  they won't need a method to delete themselves).
B. I also think that maybe A is not mandatory because we can log a warning if it was deleted before. 

B is much less work to implement, it's just checking if the file exists before trying to delete it, but it may make harder to spot when an error unrelated to the file not existing is present?
But if we go with B, why add a middleware (it would be an overkill) when we could just add the deletion of files on the error handler middleware, as it only has problems problems for now at that scenario. 
So I think we should go with A because:
- Easier to maintain
- We can be sure that the files are being deleted from the middleware, so if error it's not an `ENOENT`
- Future proof
- It isn't that much work (B is the lazy way 😝)
----

> Adding file removal to every single request in API v2 isn't the safest approach

Yes, that's right, changing only to multer uploads.

> Maybe i saw a case where a request/route might want to disable the behaviour?

@kirrg001 Optional variable in the res object, like `res.disableUploadClear` that decides if register the `res.on` events?

Also I'll be doing a unit test to verify it's working everything as intended once we figure out the last details.

kirrg001 commented:
@gargol Do you need help with reviewing? Let me know :)

gargol commented:
Thanks @kirrg001 ;) Was planning to do review and testing after lunch today. Will reach out when there are more specific questions.

gargol commented on core/server/web/api/v2/admin/app.js:
```diff
@@ -27,7 +27,6 @@ module.exports = function setupApiApp() {
 
     // Admin API shouldn't be cached
     apiApp.use(shared.middlewares.cacheControl('private'));
-
```
This is probably a miss-commited change :thinking: No need to change formatting for this file

gargol commented on core/server/web/api/v2/admin/uploadHelper.js:
```diff
@@ -0,0 +1,27 @@
+const config = require('../../../../config');
+
+const multer = {
+    clear: config.get('upload-clear') || true,
```
```suggestion
    clear: config.get('uploadClear') || true,
```

Config files are following camelCase naming, would be good to keep the convention intact :+1: So something like 'uploadClear' would be better

gargol commented on core/server/web/api/v2/admin/uploadHelper.js:
```diff
@@ -0,0 +1,27 @@
+    multer.instance = multerModule;
+};
+
+const single = type => (req, res, next) => {
```
The `type` name here is a bit confusing because `filename` is usually passed in :)

gargol commented on core/server/web/api/v2/admin/routes.js:
```diff
@@ -4,13 +4,14 @@ const multer = require('multer');
-const tmpdir = os.tmpdir;
-const upload = multer({dest: tmpdir()});
+
+upload.multerInstance(multer({dest: os.tmpdir()}));
```
Think if mutler is going to be used only once and specifically to initialize the uploadHelper, it would make sense to move this whole initialization logic into helper :thinking:

notanengineercom replied:
Yes, the wrapper was just an inital thought, I tried to change as little as possible on the existing files of Ghost just in case! Great then, I’ll just pass os.tmpdir() as an argument.

gargol commented on core/server/web/api/v2/admin/uploadHelper.js:
```diff
@@ -0,0 +1,27 @@
+const shared = require('../../../shared');
```
If we move mutler initialization logic into this module, then it's interface could be used just like a regular middleware and it can be moved to the rest of middlewares in `/shared/middlewares/upload.js`

notanengineercom replied:
Yes, sounds good! So we combine uploadHelper.js with upload-clear.js in upload.js and we have one file for the whole logic! Awesome😁

gargol commented:
@notanengineercom left couple notes :wink: think we are getting closer to merge these changes in. super excited to not have to clean my `/tmp` every week :laughing:

kirrg001 commented:
@gargol I'd suggest if the code is not in a mergable state by next Monday, you should finish up the PR and push to this branch. Otherwise the PR will be open for a couple of weeks longer 🙂

gargol commented:
Sure @kirrg001,  @notanengineercom [mentioned](https://github.com/TryGhost/Ghost/pull/10274#issuecomment-455972689) he plans to work on this feature this week but otherwise will try and finish it up next "bug fixing Monday"

notanengineercom commented:
I think it’ll be ready to merge before friday @kirrg001, @gargol. I agree that refreshing the mind is time consuming😁 thanks for the feedback!

notanengineercom commented:
Sorry for the delay, I didn't have much time until today. I just removed all the references for temp files removal and found that when uploading a zip file (which I didn't know was possible)  in the `Import content` form in the labs section, the same as when uploading a theme happens (the unzipped file stays in the temp folder on failure). So I changed a bit the [cleanUp](https://github.com/TryGhost/Ghost/pull/10274/commits/fe95380e1e3e0c7a71f752988129e04a50383b98#diff-31e1b8c3e7ffbebafcd57a2585f75f7bR104) function(in **core/server/data/importer/index.js**), and called it on every error caused inside the importer. 
For some strange reason, I couldn't get to remove the unzipped file with _.each() inside cleanUp. It would always omit the function for me! I'm not sure if I was overseeing something or doing something silly, I don't know. That's why I changed it to forEach() and used _const_ instead of _var_.

I don't see any test files regarding uploads to Ghost, so I guess we don't need them??
I manually uploaded all possible files to Ghost, and checked in temp folder to see that no files were left there, and in the console there were no errors / warnings about the files removal. 

If there's nothing else to change I'd say the PR is ready!

gargol commented on core/server/data/importer/index.js:
```diff
@@ -102,22 +102,18 @@ _.extend(ImportManager.prototype, {
      */
     cleanUp: function () {
-        var filesToDelete = this.filesToDelete;
-        return function (result) {
```
Looks like `cleanUp` wasn't functioning properly because it was returning a callable function for some reason... which would have to be used as `cleanUp()(result)`.  Your change seem to make sense :+1:

gargol commented on core/server/data/importer/index.js:
```diff
@@ -293,9 +294,7 @@ _.extend(ImportManager.prototype, {
         var self = this,
             ext = path.extname(file.name).toLowerCase();
-
-        this.filesToDelete.push(file.path);
```
Why was this changed to an empty array and not something like `self.filesToDelete = [file.path];` ? :thinking:

gargol commented:
@notanengineercom fyi, found couple minor issues and doing some more changes to this PR as we wanted to push it out tomorrow :wink:

gargol commented:
@notanengineercom so we avoid merge conflicts and have things in one place, can you give me permission to push to your branch? thanks :)

notanengineercom commented:
@gargol I just pushed the changes to the importer file, and added you to the branch!

gargol commented:
@notanengineercom tried pushing but doesn't seem to work. Could you double check please :)

notanengineercom commented:
@gargol I think you have to accept a link sent to your email, I can see `Awaiting gargol’s response`
Or use this [link](https://github.com/notanengineercom/Ghost/invitations) maybe?

gargol commented on core/server/web/api/v0.1/routes.js:
```diff
@@ -11,8 +11,7 @@ const cors = require('../../shared/middlewares/api/cors');
 // Handling uploads & imports
-const tmpdir = require('os').tmpdir;
-const upload = require('multer')({dest: tmpdir()});
+const upload = require('../../shared/middlewares/upload');
```
Had to enable this middleware for v0.1 because themes controller is still shared as well as importer service

notanengineercom commented:
@gargol, I'm about to push a commit to gscan, I think it's an easier solution as proposed before.

gargol commented:
sure @notanengineercom I'll try giving it a review next Monday ;)

notanengineercom commented:
Ohh haha, I thought it had to be done for this PR! It's a really easy and quick fix, but I think it's better than making a function to export from the module itself! We could add the theme deletion even to this PR :D @gargol

gargol commented:
Let's keep this PR small :) There's no hurry, just wanted to be done with 'middleware' part as it has been open for over a month. In general, it's best doing small but precise changes, so they are easy to review and the impact is not as big.

gargol commented:
Had a play with whole PR yesterday, and thought it would be ready for release today, but decided to take a step back and aim for next week release again. The main reasons are: there is more manual testing that is needed with API v01 and I think it's best if we break this PR into 2 smaller PRs:
1. The middleware itself being enabled on both APIs and with removed `fs.remove` usages in API v2 related code (current state seemed fine).
2. Changes done in `core/server/data/importer/index.js` are valid improvements, but don't directly touch the 1st bit. So would meke sense to PR/test separately :) 

For 1st part I think we can also remove configuration that allows disabling this cleanup. After thinking about it and playing with code, haven't found good reason to have it there (which seemd like a good idea in the beggining).

@notanengineercom let me know if you'd have time to do this cleanup by next Monday :wink: and thanks for all your work!

notanengineercom commented:
Hey @gargol, sure no problems!
I just have a few questions. 
Why did we add the middleware to v0.1? I thought we wouldn’t add it because as it’s not being developed anymore. 

So if we decide go further with v0.1, we have to delete the manual references to the removal of the files on the temp directory in the v0.1 middlewares, right? 

And last question, we would have 3 different PRs to fix this issue right? This PR with your point 1, other one with the importer changes, and the last one that applies the gscan fix. 

And sure, I’ll have it done by tomorrow night as it’s basically just separating code and changing a few bits! If you or any Ghost member have more feedback/suggestions regarding the code and structure let me know😁😁

gargol commented:
@notanengineercom 
> Why did we add the middleware to v0.1? I thought we wouldn’t add it because as it’s not being developed anymore.

Even though v0.1 is not developed it would be wrong introducing behaviour that is worse than the current one (a part of `v0.1/themes.js` was removed, which would make users of v0.1 get even more trash in their `/tmp` :grimacing:)

> So if we decide go further with v0.1, we have to delete the manual references to the removal of the files on the temp directory in the v0.1 middlewares, right?

Would leave this as is if those are stricly parts that v0.1 uses. It should not cause problem if file is already removed. Just tyring keeping the amount of changes in v0.1 codebase to minimum, but without making it worse :)

> And last question, we would have 3 different PRs to fix this issue right? This PR with your point 1, other one with the importer changes, and the last one that applies the gscan fix.

After having a small internal chat. We decided to do a more positive approach here: merge these changes tonight, so there would be just the `gsan` PR left. This way we could improve current changes in smaller interations and your contribution sees the production faster :tada: 

So would suggest to leave the code as is for now, and we could add some improvements (e.g. a unit test for new upload middleware) as you have time :+1:

notanengineercom commented:
Oh sure, I didn’t think about it! I sometimes forget about v0.1 as a standalone Ghost install! Even with going trought the v0.1 folders, I just think it’s legacy code as base for v2!
And that’s great news! As it is right now, it should work on both versions! I can try to do some unit test for the upload middleware, but it wouldn’t be added on this PR, right?

gargol commented:
@notanengineercom yes right, the unit test can come in as a follow-up PR ;)

gargol commented:
@notanengineercom it landed in master :tada:

notanengineercom commented:
Hooray🙌🏼🙌🏼
Thanks for letting me contribute to this awesome software, it really makes me proud😊. I’ll try to always be a helping hand if needed for Ghost👻
Thanks for the feedback/improvements and sorry fot the delay😜@gargol @kirrg001
